### DISCUSSION TITLE
regex_automata missing replace?

### DISCUSSION THREAD

**Discussion by VorpalBlade** [rust-lang/regex#1166] [Q&A]
# What I'm trying to accomplish

I'm interested in implementing a user-configurable set of regex replacements on some input I will process. It would be quite similar to a multi-expression sed:

```bash
sed -e 's/some match with \(capture .*\) groups/replacement with references \1/g' \
    -e 's/another match/another replacement with references/' \
    ...
```

regex has replace (using `$1` etc for capture groups), but not in RegexSet (and RegexSet doesn't support captures like that). Instead the documentation directed me towards regex-automata, but that doesn't seem to have replacements either. I could implement something myself of course on top of `Regex::new_many`, but I had hoped there would at least be a parser in regex-automata for replacement strings, or some low level support for replacements at all.

It also isn't clear to me how to implement something like sed's g-modifier (or rather lack of modifier), that is, how to make a pattern match at most once per line. While not essential, it would be nice to have for familiarity with users familiar with sed (which I expect my target audience to be, this is a tool for sysadmins / Unix greybeards, and yes I count myself to the latter group). It seems I would have to scan the ranges between matches for newlines and possibly discard matches if they are on the same line, keeping track of which pattern those matches originated from so that different patterns can still match on the same line.

# Questions

So, what are my questions then?

* Is there a better built in way to do this that I missed?
* Would it be possible to expose the parser for the regex replacement syntax (I don't like reinventing the wheel, especially for things that are not core functionality of what I'm making)?
* Would this be something you are interested in supporting fully on your end in the future? Or is that out of scope for the regex-automata API?
* What is the best way to do something like sed's "at most one match per line, per pattern"?

**Comment by BurntSushi** [comment]
Yes! It exists! The "lowest level" variant of it is in the [`regex-automata::util::interpolate`](https://docs.rs/regex-automata/latest/regex_automata/util/interpolate/index.html) module. And then there are convenient APIs for it on `Captures`, for example, [`Captures::interpolate_string`](https://docs.rs/regex-automata/latest/regex_automata/util/captures/struct.Captures.html#method.interpolate_string).

It doesn't exist on `regex-automata::meta::Regex` because the higher level API is a bit of a mouthful. That's really better suited for the facade in `regex` proper. And since a `meta::Regex` doesn't have a separate `&[u8]` and `&str` APIs, it would be even more annoying to support as it exists in `regex`.

> It also isn't clear to me how to implement something like sed's g-modifier (or rather lack of modifier), that is, how to make a pattern match at most once per line.

Have you looked at the implementation of `replacen` in the `regex` crate? It is not much code. Other than the interpolation bits and a straight-forward optimization or two, it really is just using `regex.find_iter(..)`. You should be able to write it yourself without too much trouble. With an iterator, you can control precisely how many matches to replace.

The multi-pattern case could possibly complicate this somewhat, but it really depends on the semantics you want. You might not be able to use a multi-regex at all for example if the semantics are "do replacement for regex `1` on line `N`, then replace for regex `2` on line `N`, and so on." The way to think about a multi-regex is as-if it were a single regex like so: `regex1|regex2|...|regexN`. If you get a match for `regex2` even if `regexN` would match at the same range, the match for `regexN` won't be reported. You _could_ enable overlapping matches (which is what `RegexSet` does) via `MatchKind::All`, but then the match semantics reported to users is likely to be different than what they expect because _every_ match will be reported. e.g., `a+` will match `aaa` 3 times.

> Would it be possible to expose the parser for the regex replacement syntax (I don't like reinventing the wheel, especially for things that are not core functionality of what I'm making)?

I would perhaps caution you that if the primary purpose of your CLI tool is to do replacements, then it might actually be a good idea to provide your own syntax. Because the syntax supported by `regex-automata`/`regex` is pretty barebones and it does have some non-ideal parts. I somewhat regret, for example, using it as-is for ripgrep's `-r/--replace` flag. But that's your call.

**Reply by VorpalBlade** [reply]
Thank you, I had missed that. I was searching for "replace" on docs.rs, not interpolate. Unfortunately docs.rs search really isn't very smart, nor full text.

The primary purpose of the tool I'm currently designing is to manage installed packages and config files (I'm making a cross-distro version of a existing Arch Linux specific tool (aconfmgr) that I really like). That existing tool is written in shell script and makes many assumptions that wouldn't carry over to for example Debian. It also doesn't seem well maintained any longer. 

The core idea is that it manages the entire system configuration (not unlike NixOS does) but unlike NixOS you can save your system to a file (saves the difference between your existing config and what is on the system) as well as apply your config to the system (install/uninstall packages, patch config files etc). As it is scriptable you can make system specific bits etc. This is useful for someone who has entirely too many desktops and laptops to manage.

The specific operation I'm considering here is taking a package default settings file and describing how it should differ. So it is an important part, but the primary purpose of the tool.
